eclipse-glsp/glsp-client
# Improve handling of change bounds in several areas

#### What it does

Improve handling of change bounds

- Introduce change bounds manager to centralize bounds-related services
-- Bounds changes through position snapping and movement restriction
-- Validation for size and position of an element
-- Customizable methods for when to use move and resize options

- Introduce change bounds tracker for moves and resizes
-- Tracker calculates move on diagram and calculates move and resizes
-- Tracker supports options on which parts of the process are applied

- Provide moveable wrappers for resize and routing handles

Fixes https://github.com/eclipse-glsp/glsp/issues/1337

- Extend current resize capabilities
-- Introduce mode for symmetric resize
-- Introduce one-dimensional resize on top, right, bottom and left side

Fixes https://github.com/eclipse-glsp/glsp/issues/1338
Fixes https://github.com/eclipse-glsp/glsp/issues/1339

- Fix elements moving during resizing when hitting minimum bounds
-- Store calculated minimum size from layouter in element
-- Adapt resize so we do not produce invalid sized bounds

Fixes https://github.com/eclipse-glsp/glsp/issues/1340

Minor:
- Ensure we get proper cursor feedback when hovering over resize handle
- Add additional convenience functions
- Add origin viewport command for convenience

**_Please note this PR builds upon https://github.com/eclipse-glsp/glsp-client/pull/343 and should only be merged afterwards!_**

#### How to test

- Double check that everything behaves as before (minus the bugs)
- Symmetric resize can be tested using the Ctrl Key
- Additional resize handles can be tested by adding the locations to the ChangeBoundsManager. The corners are kept for backwards-compatibilty.
- General sizing can be tested more easily with the debug bounds module from https://github.com/eclipse-glsp/glsp-client/pull/343

#### Follow-ups



#### Changelog




As soon as we agree on the PR, I'll write a paragraph about the changes ;-)
